### ISSUE TITLE
Forbid nesting a Track into a Track or automatically insert a Stack in between

### ISSUE THREAD

**Issue by Tilix4** [PixarAnimationStudios/OpenTimelineIO#1171]
## Feature Request

Proposed Change to existing Behavior

## Description

Appending a `Track` into a `Track` makes no sense if a `Stack` isn't inserted as a parent of an appended `Track`.

We could either decide to forbid appending a `Track` to a `Track` or automatically insert a `Stack` in between.

## Context

```py
from opentimelineio.schema import Track

track = Track()
track.append(Track())  # This shouldn't be allowed

track.append(Stack(Track()))  # This could be automatically done
```

**Comment by ssteinbach** [comment]
Why should inserting a track inside a track be an error?  We allow nesting both tracks inside tracks and stacks inside stacks.

**Comment by Tilix4** [comment]
Yes I know you allow it, I'm just raising that it doesn't make sense to me, for the Track > Track for sure, not sure about the Stack > Stack.

How would you represent into an editing software to have a Track inside a Track? In an architecture schema?

**Comment by ssteinbach** [comment]
This has examples of nested compositions:
https://opentimelineio.readthedocs.io/en/latest/tutorials/otio-timeline-structure.html

I don't see one on there for a track with a nested track specifically, but we could add that.  A track contains a sequence of items ordered in time such that one follows the next.  One of those items could itself be a track containing more items sequentially ordered in time.  

Some examples from third party NLEs:

Documentation for premiere pro on nested sequences:
https://helpx.adobe.com/premiere-pro/how-to/premiere-pro-nesting-sequence-cc.html

Looks like FCPX calls them 'compound clips':
https://support.apple.com/guide/final-cut-pro/intro-to-compound-clips-verbbd3496b/mac

**Comment by ssteinbach** [comment]
Here is a relevant unit test:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/b566b316b2fa4cb9254f0b5d4af4ed636513d0d3/tests/test_composition.py#L1144

**Comment by meshula** [comment]
Media composer calls them "single layer submaster".

**Comment by ssteinbach** [comment]
Ah yes, here: https://resources.avid.com/SupportFiles/attach/MCEffects.pdf

search for "submaster"

**Comment by Tilix4** [comment]
I completely understand what you mean, but all your examples show a `Stack` having several `Track`s with their clips and put inside a "bigger" `Track`. That's my original point, saying: "I want`Track`s into a `Track`" would actually mean "I want a `Track` to contain a `Stack` with numerous `Track`s in it".

That being said, we could consider a `Track` is a `Stack` with only one single `Track` in it, but to me, it's open to interpretation and then error-prone.

After some thoughts, storing a `Stack` into a `Stack` has the same conceptual problem, as a `Stack` refers to something with layers (= `Track`) then being able to have a `Clip` or a `Stack` directly in a `Stack` is quite the same as wanting to append a `Clip` to another `Clip`.

If you append two `Clip`s to a `Stack`, which one should be displayed? By order of appending? Is this supposed to represent layers, then `Track`s? If yes, why not having actual `Track` between the `Stack` and the `Clip`? I may want to mean: "If two clips are in a stack, they are supposed to be mixed in equality, every pixel takes 50% of each clip, or we addition the pixels values as it does for sound...", and then I must have a "rule" of compositing for video clips into a `Stack`. Forcing to have a `Track` inside a `Stack` before adding any entity makes it, IMO, stronger.

**Comment by ssteinbach** [comment]
> That's my original point, saying: "I want Tracks into a Track" would actually mean "I want a Track to contain a Stack with numerous Tracks in it".
These are two separate concepts to us.  A stack and a track both compose their children in _time_.   Having a stack with three clips in it rather than three tracks each with one clip would render to the same thing but could mean something different organizationally.  You could also have two clips and a track containing three more clips in order.  Similarly, a track could contain three clips, or three tracks each that contain one clip, or three stacks that contain a track that contain a clip.  These all could mean different things depending on a workflow but flatten/render to the same result.  Each intermediate scope could have effects and metadata placed on it, for example.

> If you append two Clips to a Stack, which one should be displayed? 
This is well defined.  Unfortunately, when we ported to C++ we lost some of the verbose documentation that was embedded in the pydocs.  I think there is an issue to bring the documentation back?  @reinecke might know.   In any case, this explains the compositing order of a stack:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/cc78a7c5e9b10808d524edd00a09f8a11b21ce1a/opentimelineio/schema/stack.py#L32

It would be good to ensure that this is somewhere in RTD as well, but I'm a bit short on time.

> Is this supposed to represent layers, then Tracks? 
Stacks and Tracks both can contain child tracks, stacks, clips and gaps.  Tracks can additionally contain transitions.  Each of these objects has a temporal coordinate system and a well defined (hopefully) way of interpreting further nesting.  So the contents of a stack are objects with a temporal coordinate system whose starting coordinate lines up with the start of the track.  They could be tracks, with further sequencing, or just clips.

Might be easier to discuss on slack or in a TSC meeting, in real time.

**Comment by meshula** [comment]
I think the summary you came up with, "stacks are temporarily coincident (i.e. things that start playing at the same time)" and "tracks are temporally adjacent (i.e. things that play one after the other, with a sneaky twist for blends)" is a really nice way to summarize it. This conception allows thinking more easily about a timeline composition as a flow graph. The choice of the terms "track" in particular in retrospect is probably not as good a term as "sequence". Tracks in OTIO coincidentally look an awful lot like a track in an NLE, but the term doesn't necessarily frame the data structure intuitively, or correspond to a track in an NLE. "OTIO Stacks group things that start at the same time, Tracks group things that start one after the other" really helps.

**Comment by JeanChristopheMorinPerso** [comment]
> I think there is an issue to bring the documentation back? @reinecke might know.

There is one I believe yes. BTW, I've been doing work on the python docs, and bringing back the docs that was lost in translation is part of the things I'm working on.

**Comment by ssteinbach** [comment]
Awesome!  That will be a huge help!  Thanks!

**Comment by Tilix4** [comment]
> "stacks are temporarily coincident (i.e. things that start playing at the same time)" and "tracks are temporally adjacent (i.e. things that play one after the other, with a sneaky twist for blends)"

> "OTIO Stacks group things that start at the same time, Tracks group things that start one after the other"

I agree this helps a lot and completely makes sense. My first wonder was because I was thinking about it spatially, like a bookshelf, where the Stack is the support and the Track is the shelf, then putting a shelf on a shelf doesn't make sense and may require a support first. I came up with that because that's how NLE editors represent the editing. 
Actually, putting it in a time based approach makes more sense. Thinking about Track like a music track, with events following each others is helpful. 
I still think this gives too much freedom to implementation in different softwares and then weakens the standard, but I don't know if this is something we could completely get rid of.

I agree the doc lacks of description, use cases and examples. But this is something which takes time and the whole thing is still young.

No problem to further discuss it on Slack or in live.

**Comment by jminor** [comment]
It might be helpful to share a little more background on the design choices that lead to how OTIO is now. We considered several different options on how the composition structure works when designing Timeline, Stack, and Track.

In one extreme which is most similar to how NLE user interfaces are presented, Timeline and Stack could be the same thing, (e.g. a Timeline has a list of Tracks) and the schema could be limited so that Tracks only ever appear as direct children of a Timeline. This would mean that any nesting would be accomplished by adding a Timeline as an item within a Track, and the nested Timeline would always have a list of Tracks (starting with just 1). In the terms of our class hierarchy, Timeline would then be a subclass of Composable (and an Item, I think).

In the other extreme, which is what we chose for OTIO, the notion of Track and Stack are kept separate, and each of those two classes is only responsible for the way their child items are composed. Tracks arrange their children sequentially in time (in fact we renamed "Sequence" to "Track" early on). Stacks arrange their children vertically for compositing (video), or summation (audio). The only functional difference between the two is in the duration() computation, and the range_of_child() computation.

We chose to go with the 2nd approach on the basis that there would be other scenarios where these concepts would be useful to reuse, and that the composition logic would be simpler to implement and reason about. We saw the node graph within RV as a good example of this. RV has "Sequence" and "Stack" objects in their graph.

In practice, we've seen cases with EDL-based tools, and some of our in-house playback tools where nesting is available but there is only 1 media track, and so nesting a Track within a Track comes up there - though it could be accommodated in the other design as well. We have also considered situations like multi-cam clips, or representations of the streams within a single media file, where these choices could be reconsidered.

---
Joshua Minor (he-him)
OpenTimelineIO Technical Steering Committee Chairperson
***@***.***
